ext/sqlite3/database.c

```diff
@@ -259,6 +261,45 @@ busy_handler(int argc, VALUE *argv, VALUE self)
     return self;
 }
 
+static int
+rb_sqlite3_statement_timeout(void *context)
+{
+    sqlite3RubyPtr ctx = (sqlite3RubyPtr)context;
+    struct timespec currentTime;
+    clock_gettime(CLOCK_MONOTONIC, &currentTime);
+
+    if (ctx->stmt_deadline == 0) {
+        ctx->stmt_deadline = currentTime.tv_sec * 1e9 + currentTime.tv_nsec + (long long)ctx->stmt_timeout * 1e6;
```

Comment:
You could just make `stmt_deadline` a `timespec`.

Reply:
Genuine question:

Do you find this code "better" (in whatever ways one might access C code in this library)?
```c
static int
rb_sqlite3_statement_timeout(void *context)
{
    sqlite3RubyPtr ctx = (sqlite3RubyPtr)context;
    struct timespec currentTime;
    clock_gettime(CLOCK_MONOTONIC, &currentTime);

    if (ctx->stmt_deadline.tv_sec == 0) {
        // Set stmt_deadline if not already set
        time_t timeout_sec = ctx->stmt_timeout / 1000;
        ctx->stmt_deadline.tv_sec = currentTime.tv_sec + timeout_sec;
        ctx->stmt_deadline.tv_nsec = currentTime.tv_nsec;
    } else {
        // Calculate time difference directly using timespec
        struct timespec timeDiff;
        timeDiff.tv_sec = ctx->stmt_deadline.tv_sec - currentTime.tv_sec;
        timeDiff.tv_nsec = ctx->stmt_deadline.tv_nsec - currentTime.tv_nsec;

        // Normalize time difference
        while (timeDiff.tv_nsec < 0) {
            timeDiff.tv_sec--;
            timeDiff.tv_nsec += 1000000000;
        }

        // Check if time difference is negative
        if (timeDiff.tv_sec < 0 || (timeDiff.tv_sec == 0 && timeDiff.tv_nsec <= 0)) {
            return 1;
        }
    }

    return 0;
}
```

I find the normalization and final check more confusing than the math in the current version :shrug:

Reply:
What I was suggesting to do was:

  - Make `ctx->stmt_deadline` a `struct timespec`.
  - Implement a `timespecadd` function, IIRC it is included on some system, but Linux might not have it. But basically it's similar to your thing.
  - Then continue with your initial algo, which is to compute the deadline when the query is triggered
  - Then in the callback, get current time and compare it with the deadline.

Everything can be made quite easy to grasp if the gritty details of how to add and compare `struct timespec` is moved into sub functions.

Reply:
+1 on @byroot's suggestion. I like the idea of storing the `timespec` in the `sqlite3RubyPtr` struct. Then we could use the [timespec functions](https://nxmnpg.lemoda.net/3/timespecadd) to do this math (or implement our own on platforms that don't have it).

Reply:
@tenderlove: I implemented things using `timespec`s in [bd0267e](https://github.com/sparklemotion/sqlite3-ruby/pull/463/commits/bd0267e2923a19a247c5dee0745dbfd6e5b8bbe2), but for some reason CRuby builds fail trying to `#include <timespec.h>`. I have no idea what might the problem there, as the error message that the file doesn't exist, but is certainly does exist. Any help at all would be greatly appreciated.

Reply:
Ignore my comment.

Reply:
@fractaledmind You need to add timespec.h to the gemspec so it's packaged into the gem!

Reply:
Thank you for the tip! Fixed with one randomly failing action in CI. Should be good to go.